server/data/authClient.ts

```diff
@@ -22,24 +24,54 @@ const agentOptions = {
 
 const keepaliveAgent = apiUrl.startsWith('https') ? new HttpsAgent(agentOptions) : new Agent(agentOptions)
 
-async function getSystemClientToken(username?: string) {
-  const clientToken = generateOauthClientToken(config.apis.oauth2.systemClientId, config.apis.oauth2.systemClientSecret)
+async function getSystemClientToken(tokenStore: TokenStore, username?: string) {
+  const key = username || '%ANONYMOUS%'
+  const token = await tokenStore.getToken(key)
 
-  const oauthRequest = username
-    ? querystring.stringify({ grant_type: 'client_credentials', username })
-    : querystring.stringify({ grant_type: 'client_credentials' })
+  if (token) {
+    return token
+  }
 
-  logger.info(
-    `Oauth request '${oauthRequest}' for client id '${config.apis.oauth2.apiClientId}' and user '${username}'`
+  const clientTokenHeader = generateOauthClientToken(
+    config.apis.oauth2.systemClientId,
+    config.apis.oauth2.systemClientSecret
   )
 
-  const result = await superagent
+  const oauthRequest = username ? { grant_type: 'client_credentials', username } : { grant_type: 'client_credentials' }
+  const newToken = await oauthTokenRequest(clientTokenHeader, oauthRequest)
+
+  logger.info(`Oauth request '${oauthRequest}' for client id '${config.apis.oauth2.apiClientId}' and user '${key}'`)
+
+  await tokenStore.setToken(key, newToken.token, newToken.expiresIn - 60)
+  return newToken.token
+}
+
+const oauthTokenRequest = async (clientTokenHeader, oauthRequest) => {
+  const oauthResult = await getOauthToken(clientTokenHeader, oauthRequest)
+  logger.info(`Oauth request for grant type '${oauthRequest.grant_type}', result status: ${oauthResult.status}`)
+
+  return parseOauthTokens(oauthResult)
+}
+
+const getOauthToken = (oauthClientToken, requestSpec) => {
+  const oauthRequest = querystring.stringify(requestSpec)
+  const handleError = buildErrorHandler('OAuth')
```

Comment:
Can we get rid of this error handling bit and just leave the error handling as it was?

Reply:
removed